include/gul17/join_split.h

```diff
@@ -324,6 +362,103 @@ join(const StringContainer& parts, std::string_view glue)
+ *                   boost. The default is zero (no preallocation).
+ *
+ * \returns the concatenated string.
+ *
+ * \tparam Iterator  A forward iterator type that dereferences to a type that can be
+ *                   passed to the ConversionFct.
+ * \tparam ConversionFct  A function object type that can be called with the dereferenced
+ *                   iterator type and returns a string type. This string type must
+ *                   support concatenation with std::string::operator+=().
+ *
+ * \since GUL version UNRELEASED
+ */
+template <typename Iterator, typename ConversionFct,
+    std::enable_if_t<std::is_invocable_v<ConversionFct, decltype(*Iterator{})>, bool>
+    = true>
+inline std::string
+join(Iterator begin, Iterator end, std::string_view glue, ConversionFct to_string,
+     std::size_t prealloc = 0)
+{
+    std::string result;
+
+    if (begin == end)
+        return result; // Return an empty string
+
+    result.reserve(prealloc);
+    result += to_string(*begin);
+
+    // Iterate over all but the first element of the range
+    for (auto it = std::next(begin); it != end; ++it)
+    {
+        result += glue;
+        result += to_string(*it);
+    }
+
+    return result;
+}
+
+/**
+ * Concatenate the strings resulting from calling the given function on each element
+ * in a range, placing a delimiter between them.
+ *
+ * \code{.cpp}
+ * std::vector<int> parts = { 1, 2, 3 };
+ * std::string result = join(parts, "-", [](int i) { return std::to_string(i); });
+ * assert(result == "1-2-3");
+ * \endcode
+ *
+ * This algorithm iterates exactly once over the range. To speed up complex joining
+ * operations, string memory can be preallocated via the last parameter.
+ *
+ * \param container  Container of input elements
+ * \param glue       String to be put between the string-converted elements
+ * \param to_string  Function object to convert the container elements to strings
+ * \param prealloc   Number of bytes to preallocate for the result string. Providing this
+ *                   parameter is not necessary, but it may provide a tiny performance
+ *                   boost. The default is zero (no preallocation).
+ *
+ * \returns the concatenated string.
+ *
+ * \tparam Container   A container type that provides a cbegin() and cend() member
+ *                     function returning forward iterators.
+ * \tparam ConversionFct  A function object type that can be called with the dereferenced
+ *                     iterator type and returns a string type. This string type must
+ *                     support concatenation with std::string::operator+=().
+ *
+ * \since GUL version UNRELEASED
+ */
+template <typename Container, typename ConversionFct,
+    std::enable_if_t<std::is_invocable_v<ConversionFct,
+                     decltype(*(std::declval<Container>().cbegin()))>, bool> = true>
+inline std::string
+join(const Container& container, std::string_view glue, ConversionFct to_string,
```

Comment:
This is roughly

```c++
template <typename T>
inline RetType function(ArgType);
```

Why is that `inline`? It's automatically inline as it's an ordinary template?
What shall the `inline` express?

Reply:
This "inline" appears in a few places in GUL. I guess we started adding it because a normal function would need it if it were declared and defined in the same place in the header file. Should I remove it from the function templates in this file?